build/parseList.c

```diff
@@ -8,13 +8,99 @@
 #include "build/rpmbuild_internal.h"
 #include "debug.h"
 
+static int addLinesFromFile(rpmSpec spec, const char * const fn, rpmTagVal tag)
+{
+    int nlines = 0;
+    int flags = STRIP_COMMENTS | STRIP_TRAILINGSPACE;
+    ARGV_t argv = NULL;
+
+    int terminate = rpmExpandNumeric(tag == RPMTAG_SOURCE
+				     ? "%{?_empty_sourcelist_terminate_build}"
+				     : "%{?_empyy_patchlist_terminate_build}");
+
+    if (!terminate)
+	flags |= ALLOW_EMPTY;
+
+
+    nlines = readManifest(spec, fn, tag == RPMTAG_SOURCE ? "%sourcelist" : "%patchlist", flags, &argv, NULL);
+
+    for (ARGV_t av = argv; *av; av++) {
+	addSource(spec, 0, *av, tag);
+    }
+
+    argvFree(argv);
+
+    if (nlines <= 0) {
+	rpmlog(terminate ? RPMLOG_ERR : RPMLOG_WARNING,
+	       _("Empty %s file %s\n"),
+	       tag == RPMTAG_SOURCE ? "%sourcelist" : "%patchlist",
+	       fn);
+    }
+
+    return (nlines < 0) ? RPMRC_FAIL : RPMRC_OK;
+}
 
 int parseList(rpmSpec spec, const char *name, rpmTagVal tag)
 {
     int res = PART_ERROR;
     ARGV_t lst = NULL;
+    int rc, argc = 0;
+    int arg;
+    const char **argv = NULL;
+    poptContext optCon = NULL;
+    struct poptOption optionsTable[] = {
+	{ NULL, 'f', POPT_ARG_STRING, NULL, 'f', NULL, NULL},
+	{ 0, 0, 0, 0, 0, NULL, NULL}
+    };
+    char * file = NULL;
+
+    if ((rc = poptParseArgvString(spec->line, &argc, &argv))) {
+	rpmlog(RPMLOG_ERR, _("line %d: Error parsing %%%s: %s\n"),
+	       spec->lineNum, name, poptStrerror(rc));
+	goto exit;
+    }
+
+    optCon = poptGetContext(NULL, argc, argv, optionsTable, 0);
+    while ((arg = poptGetNextOpt(optCon)) > 0) {
+	;
+    }
+
+    if (arg < -1) {
+	rpmlog(RPMLOG_ERR, _("line %d: Bad option %s: %s\n"),
+	       spec->lineNum,
+	       poptBadOption(optCon, POPT_BADOPTION_NOALIAS),
+	       spec->line);
+	goto exit;
+    }
+
+    lst = argvNew();
+
+    for (arg = 1; arg < argc; arg++) {
+	if (rstreq(argv[arg], "-f")) {
+
+	    if (!argv[arg+1]) {
+		rpmlog(RPMLOG_ERR,
+		       _("line %d: \"%%%s -f\" requires an argument.\n"),
+		       spec->lineNum, name);
```

Comment:
We have popt for option and argument parsing, use it. Manually testing for options and their arguments doesn't cut it.

Reply:
Done.